access-widener-javaparser/src/main/java/net/fabricmc/accesswidener/javaparser/SourceAccessWidenerTransformer.java

```
@@ -0,0 +1,196 @@
+
+import net.fabricmc.accesswidener.AccessWidener;
+import net.fabricmc.accesswidener.EntryTriple;
+
+public class SourceAccessWidenerTransformer {
+	protected final AccessWidener widener;
+	protected final PrintStream warn;
+
+	public SourceAccessWidenerTransformer(AccessWidener widener, PrintStream warn) {
+		this.widener = widener;
+		this.warn = warn;
+
+		if(!widener.isSourceCompatible()) {
+			throw new UnsupportedOperationException("AccessWidener must have source compatible entries!");
+		}
+	}
+
+	public SourceAccessWidenerTransformer(AccessWidener widener) {
+		this(widener, System.err);
+	}
+
+	/**
+	 * A heuristic on whether the java file may be transformed.
+	 */
+	public boolean mayTransform(String javaFilePath) {
+		String pkg = AccessWidener.getPackage(javaFilePath);
+		return this.widener.getPackages().contains(pkg);
+	}
+
+	public boolean transform(CompilationUnit unit) {
+		boolean transformed = false;
+
+		for(TypeDeclaration<?> type : unit.findAll(TypeDeclaration.class)) {
+			String qualifiedName = getFullyQualifiedName(type).orElse(null);
+
+			if(qualifiedName == null) {
+				System.err.println("Unable to find internal name of local class in " + unit
+						.getPrimaryType()
+						.flatMap(SourceAccessWidenerTransformer::getFullyQualifiedName) + ".java");
+				return false;
+			}
+
+			if(this.widener.getTargets().contains(qualifiedName)) {
+				return false;
+			}
+
+			boolean isInterface = type.isClassOrInterfaceDeclaration() && type
+					.asClassOrInterfaceDeclaration()
+					.isInterface() || type.isAnnotationDeclaration();
+
+			String internalName = qualifiedName.replace('.', '/');
+			NodeList<Modifier> classModifiers = type.getModifiers();
+			int classFlags = SourceUtil.toClassFlags(classModifiers, type);
+			int newFlags = this.widener.getClassAccess(internalName).apply(classFlags, internalName, classFlags);
+			SourceUtil.fromFlags(classModifiers, newFlags, false);
+
+			// todo add unsealing support when javaparser supports it
```

Does it get this far, can it know if a class is sealed and thow an UOE?

---

From what I see online, it will throw when u try to parse the CompileUnit (it'll give u a parse error in the result object)

---

Ok great 👍 as long as it doesn’t silently fail.